numba/np/ufunc/deviceufunc.py

```diff
@@ -105,7 +105,10 @@ def _fill_argtypes(self):
         """
         for i, ary in enumerate(self.arrays):
             if ary is not None:
-                self.argtypes[i] = np.asarray(ary).dtype
+                dtype = getattr(ary, 'dtype')
+                if dtype is None:
+                    dtype = np.asarray(ary).dtype
+                self.argtypes[i] = dtype
```

Comment:
I see. The `np.asarray` is forcing a transfer to the host while we only care about getting the `.dtype`.